src/eve/datamodels.py

```diff
@@ -0,0 +1,1008 @@
+        return type_hint[tuple(type_params_map[tp] for tp in type_hint.__parameters__)], True
+    else:
+        return type_hint, False
+
+
+def _make_counting_attr_from_attribute(
+    field_attrib: attr.Attribute, *, include_type: bool = False, **kwargs: Any
+) -> Any:  # attr.s lies on purpose in some typings
+    members = [
+        "default",
+        "validator",
+        "repr",
+        "eq",
+        "order",
+        "hash",
+        "init",
+        "metadata",
+        "converter",
+        "kw_only",
+        "on_setattr",
+    ]
+    if include_type:
+        members.append("type")
+
+    return attr.ib(**{key: getattr(field_attrib, key) for key in members}, **kwargs)  # type: ignore  # too hard for mypy
+
+
+def _make_dataclass_field_from_attr(field_attrib: attr.Attribute) -> dataclasses.Field:
+    MISSING = getattr(dataclasses, "MISSING", NOTHING)
+    default = MISSING
+    default_factory = MISSING
+    if isinstance(field_attrib.default, attr.Factory):  # type: ignore  # attr.s lies on purpose in some typings
+        default_factory = field_attrib.default.factory  # type: ignore  # attr.s lies on purpose in some typings
+    elif field_attrib.default is not attr.NOTHING:
+        default = field_attrib.default
+
+    assert field_attrib.eq == field_attrib.order  # dataclasses.compare == (attr.eq and attr.order)
+
+    dataclasses_field = dataclasses.Field(  # type: ignore  # dataclasses.Field signature seems invisible to mypy
+        default=default,
+        default_factory=default_factory,
+        init=field_attrib.init,
+        repr=field_attrib.repr if not callable(field_attrib.repr) else None,
+        hash=field_attrib.hash,
+        compare=field_attrib.eq,
+        metadata=field_attrib.metadata,
+    )
+    dataclasses_field.name = field_attrib.name
+    assert field_attrib.type is not None
+    dataclasses_field.type = field_attrib.type
+
+    return dataclasses_field
+
+
+def _make_non_instantiable_init() -> Callable[..., None]:
+    def __init__(self: DataModelLike, *args: Any, **kwargs: Any) -> None:
+        raise TypeError(f"Trying to instantiate '{type(self).__name__}' abstract class.")
+
+    return __init__
+
+
+def _make_post_init(has_post_init: bool) -> Callable[[DataModelLike], None]:
+    if has_post_init:
+
+        def __attrs_post_init__(self: DataModelLike) -> None:
+            if attr._config._run_validators is True:  # type: ignore  # attr._config is not visible for mypy
+                for validator in type(self).__datamodel_validators__:
+                    validator.__get__(self)(self)
+                self.__post_init__()
+
+    else:
+
+        def __attrs_post_init__(self: DataModelLike) -> None:
+            if attr._config._run_validators is True:  # type: ignore  # attr._config is not visible for mypy
+                for validator in type(self).__datamodel_validators__:
+                    validator.__get__(self)(self)
+
+    return __attrs_post_init__
```

Comment:
Just for fun, didn't try if it is working and don't know if a good idea. ;)

```suggestion
def _make_post_init(has_post_init: bool) -> Callable[[DataModelLike], None]:
    def __attrs_post_init__(post_init: [Callable[], None]) -> Callable[[Callable[], None], Callable[[DataModelLike], None]]:
        def _impl(self: DataModelLike) -> None:
            if attr._config._run_validators is True:  # type: ignore  # attr._config is not visible for mypy
                for validator in type(self).__datamodel_validators__:
                    validator.__get__(self)(self)
                post_init()
        return _impl

    return __attrs_post_init__(self.__post_init_() if has_post_init else lambda: None)
```